### PR TITLE
#73

### PR DESCRIPTION
**Pull request by socketwiz**
https://github.com/t3chnoboy/amazon-product-api/issues/73

I'm not sure if this is a proper fix, but its a start in the right direction I believe...what do you think?

* Make callback and promise returns consistent.
* Update tests to pass with changes.

### CODE REVIEW

**Comment by masterT** [comment]
Maybe it could return an special AmazonResponse that has special properties such as response, results, and maybe the error. It could also have 'hasFailed', 'hasSucceeded' function. I don'exactly know. 

It could also offer a special 'dig' function to gently retrieve element in the response. [Something like in Ruby](http://ruby-doc.org/core-2.3.0/Hash.html#method-i-dig)

This way there will be consistency. What do you think? 😊

**Comment by socketwiz** [comment]
> Maybe it could return an special AmazonResponse that has special properties such as response, results, and maybe the error. 

How about something simple like:
```javascript
var amazonResponse = {
    'resp': response,
    'data': resp[method + 'Response']
};
```
then let the caller do with it whatever they like with amazonResponse?

> It could also have 'hasFailed', 'hasSucceeded' function. I don'exactly know.

Looking at  this a bit further I kind of like how error is handled already but I see I didn't handle this correctly in my initial attempt.  I need to put the `err` parameter back into the callback method the way it was before because I didn't notice that `failure()` call.  The promise already properly does a `reject()`.

> It could also offer a special 'dig' function to gently retrieve element in the response.

I'd let the caller use something like `lodash.get()`, no need to re-invent the wheel here: https://lodash.com/docs#get

Let me get this coded up and see what you guys think...

**Comment by socketwiz** [comment]
I've updated the pull request to reflect the changes discussed above.  Let me know if you would like me to change anything.

**Comment by t3chnoboy** [comment]
Thank you for doing this @socketwiz!
The changes make sense.
The only drawback is that this is a breaking change. So we would need to bump the major version
and break compatibility with the existing clients.
Any ideas how to handle this more gracefully?
Like adding a new method with a different name and making this deprecated, without breaking compatibility.

**Comment by t3chnoboy** [comment]
@masterT @socketwiz @JuanCrg90 Your thoughts? ^^^

**Comment by masterT** [comment]
Let's bump to major version, but I don't think we need to change the function names, maybe add a `CHANGELOG.md` file?

**Comment by masterT** [comment]
Good job @socketwiz 👍! I didn't know `lodash.get()` thank it can be really useful!

**Comment by JuanCrg90** [comment]
I think that it's a good idea add the Changelog file. and jump to a major version.

do we need update all the examples?

**Comment by socketwiz** [comment]
Changelog added and version bumped.

**Comment by socketwiz** [comment]
I would recommend you make a tag and release for 0.4.3 and another one for 1.0.0

**Comment by socketwiz** [comment]
> do we need update all the examples?

The examples don't do anything with the output, so they should all still work.

**Comment by t3chnoboy** [comment]
can we merge #74 before releasing v1.0.0?

File: lib/index.js [hunk 1]
```diff
@@ -16,15 +16,15 @@ var runQuery = function (credentials, method) {
     var p = new Promise(function(resolve, reject) {
       var success = function(results) {
         if (typeof cb === 'function') {
-          cb.apply(null, [null].concat(Array.prototype.slice.call(arguments)));
+          cb(null, results);
         }else{
```
Comment (masterT) [inline lib/index.js]: It think we should use a linter or something to well define the code style for this project. But yes, I think it's more sexy with space!

Comment (JuanCrg90) [inline lib/index.js]: I like the idea, @t3chnoboy @masterT  what style guide do you recommend? 

i propose the airbnb styleguide https://github.com/airbnb/javascript

**Comment by socketwiz** [comment]
If you guys want to pull in #74 I'll get to work on the linter and fix up the new tests to work with the API changes once they get merged in.

**Comment by socketwiz** [comment]
One of the airbnb requirements is:

> Unexpected var, use let or const instead

But I think to do that properly we would need to transpile the source with Babel...do we want to take it to that level or should I just convert this error to a warning for now?

**Comment by socketwiz** [comment]
Here are a couple more errors that would require an ES6 transpiler:

> Unexpected function expression                          prefer-arrow-callback
> Unexpected string concatenation                         prefer-template

**Comment by masterT** [comment]
@socketwiz PR #74 is merged!

**Comment by masterT** [comment]
What about [JavaScript Standard Style](http://standardjs.com/rules.html)?

**Comment by socketwiz** [comment]
I implemented the airbnb style throughout with the following exceptions:

- no-var was set to warning
- prefer-arrow-callback was set to warning
- prefer-template was set to warning

There were maybe 3 changes it wanted to make that I wasn't comfortable making, maybe someone more familiar with the code can make those.  I just put an eslint-disable comment on them for now.

I've also resolved the conflicts from #74 as well as fixed all of the tests, everything should be green now.

Let me know if there is anything I need to do.

**Comment by masterT** [comment]
Maybe we should discuss about the linter we want before?

**Comment by socketwiz** [comment]
> Maybe we should discuss about the linter we want before?

Thats fine, I can back it out.  Its scope creep for this ticket anyway...maybe it needs its own ticket?

What do I need to do to get this merged?

**Comment by JuanCrg90** [comment]
LGTM

**Comment by masterT** [comment]
@socketwiz
>  I created issue #77 to discuss the linter issue
> - masterT commented 2 days ago

**Comment by JuanCrg90** [comment]
Sorry i was a little busy. i have approved the changes, can we merge @masterT @t3chnoboy ?

File: CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
+## Version 1.0.0 2017-01-28
+- Updated return format so that both promises and callback now have the same structure:
+
+```javascript
+{
+  'resp': <req response>,
```
Comment (t3chnoboy) [inline CHANGELOG.md]: I don't think we want to save characters here, so `response` would be a better name.
Also, it's unclear what's `response` and what's `data`.
How about using more descriptive names and maybe adding some examples or describing the shape of those data structures?

File: CHANGELOG.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
+## Version 1.0.0 2017-01-28
```
Comment (t3chnoboy) [inline CHANGELOG.md]: Not critical, but it would be nice to add a migration guide.

File: lib/utils.js [hunk 4]
```diff
@@ -10,32 +10,42 @@ var generateSignature = function (stringToSign, awsSecret) {
 var sort = function (object) {
   var sortedObject = {};
   var keys = Object.keys(object).sort();
-  for (var i = 0; i < keys.length; i++) {
+  var i = 0;
```
Comment (t3chnoboy) [inline lib/utils.js]: Weird. Which `eslint` warning did it trigger?

File: test/amazonProductApiTest.coffee [hunk 5]
```diff
@@ -104,9 +103,8 @@ describe 'client.itemLookup(query, cb)', ->
 
     describe 'when callback is passed', ->
       it 'should return search results from amazon', (done) ->
-        client.itemLookup query, (err, results, response) ->
-          results.should.be.an.Array
-          response.should.be.an.Array
+        client.itemLookup query, (err, results) ->
+          results.should.be.an.Object
```
Comment (t3chnoboy) [inline test/amazonProductApiTest.coffee]: Would be nice to check if the object contains the required properties and also check the property types.

File: lib/utils.js (hunk 4 above)
Comment (socketwiz) [inline lib/utils.js]: http://eslint.org/docs/rules/vars-on-top

File: CHANGELOG.md (hunk 3 above)
Comment (socketwiz) [inline CHANGELOG.md]: I'll see what I can come up with.

File: CHANGELOG.md (hunk 2 above)
Comment (socketwiz) [inline CHANGELOG.md]: OK, I'll see what I can do.

File: test/amazonProductApiTest.coffee (hunk 5 above)
Comment (socketwiz) [inline test/amazonProductApiTest.coffee]: Did you have any particular properties you'd like tested for? Is there a list some where of "required" properties?

Comment (t3chnoboy) [inline test/amazonProductApiTest.coffee]: Sorry, not required, I meant the properties we defined (currently it's `resp` and `data`)

File: CHANGELOG.md (hunk 2 above)
Comment (socketwiz) [inline CHANGELOG.md]: > How about using more descriptive names

What names do you recommend?

**Comment by socketwiz** [comment]
I've addressed the comments.  If you want better names than response and data let me know what they should be, otherwise this is done...again.

File: CHANGELOG.md (hunk 2 above)
Comment (t3chnoboy) [inline CHANGELOG.md]: How about `response` and `results`?

Comment (masterT) [inline CHANGELOG.md]: It really speaks by itself!

**Comment by socketwiz** [comment]
`data` changed to `result`.  Let me know what else this thing needs.

**Comment by t3chnoboy** [comment]
@socketwiz:  @masterT and I meant the `results` array (as it is in “0.4.3”) instead of the `result` object.

**Comment by socketwiz** [comment]
> @socketwiz: @masterT and I meant the results array (as it is in “0.4.3”) instead of the result object.

Sorry, I'm not sure I'm following...what needs to be done exactly?

**Comment by t3chnoboy** [comment]
@socketwiz  In 0.4.3 we return `respObj.Items[0].Item` as `results`, which is an array that contains search results. And we return `respObj.Items` as `response`, which carries some additional information.  
 
On your branch it returns just `respObj` which is different from what we have and in my opinion is less convenient because for most cases you'd need just `results`. So I think it would be better to return the `results` array instead of `result`. Also will we miss anything if as the `response` property we return `respObj.Items` instead of the raw response (`resp`) ?

**Comment by socketwiz** [comment]
> So I think it would be better to return the results array instead of result.

I can't think of a single instance where this is all you want.  You don't have access to `TotalPages`.  This is the whole reason I made this PR to begin with because the promise only returns the results array which is kind of useless.

> Also will we miss anything if as the response property we return respObj.Items instead of the raw response (resp) ?

All we need is the meta data and the array which is exactly what my very first pr did.  Granted it did it all in one object but I digress.  Then someone said, lets return the response object and the results, so I did that, then someone said, hey lets throw in the airbnb style rules while were at it, so I did that, then someone said, well this is a a breaking change we need some documentation and don't call the objects resp and data lets call it response and results, so I did that, but now thats not good enough because the data being returned is all wrong.

I just want the promise to return the meta data as well as the array.  I really don't care how that happens.  I'm just getting tired of going round and round with this thing.  Every time I fix it, some else chimes in with some other change.  Can someone just take charge and decide what you want so we can be done with this thing?  Because I don't feel like we're any closer to a resolution from the day I made my first pull request :(

**Comment by t3chnoboy** [comment]
@socketwiz 
I'm sorry to hear that. I totally agree that most of the changes were not relevant to this PR 😔 
You didn't have to add a linter and didn't have to write the changelog but you didn't wait for us to do that. And thank you for that! 🙂 

This PR is almost good to go. The only thing I'd want, is to make it return an object that has the `results` array property (`respObj.Items[0].Item`). Because that's how this library was always working.

As for the `response` property, if `respObj.Items` contains all the necessary data (like `TotalPages`) then maybe it's fine to return just `respObj.Items`. If it's missing some necessary metadata then just return `respObj`.

So the goal is:
```javascript
client.itemSearch({
  director: 'Quentin Tarantino',
  actor: 'Samuel L. Jackson',
  searchIndex: 'DVD',
  audienceRating: 'R',
  responseGroup: 'ItemAttributes,Offers,Images'
}).then(function(results){
   // results is and Object containing two properties
   // {
   //      response: - response object with metadata (I don't know what's better in this case and asking you what's better, `respObj.Items` or `respObj`)
   //      results: - search results array (respObj.Items[0].Item)
   // }
}).catch(function(err){
  console.log(err);
});
```

**Comment by socketwiz** [comment]
> You didn't have to add a linter and didn't have to write the changelog but you didn't wait for us to do that.

I didn't really mind, I just think it was out of scope.

> This PR is almost good to go. The only thing I'd want, is to make it return an object that has the results array property (respObj.Items[0].Item). 

I think I have it now, let me know what you think.  It now returns a meta data object and a results array.  Tests are all passing.  I checked browse nodes, item search and item lookup.  Let me know if I missed something.

**Comment by t3chnoboy** [comment]
Cool! 
So now you can access `TotalPages` with the rest of metadata from the `response` and we still have the `results` array, right? If so, can you rename `result` to `results` and update the docs to reflect the changes? and then we can finally ship it 🚢 🚢 🚢

**Comment by socketwiz** [comment]
> So now you can access TotalPages with the rest of metadata from the response and we still have the results array, right?

That is correct.

> If so, can you rename result to results and update the docs to reflect the changes?

Yes good catch.

I've updated result to results in the library, tests and supporting documentation.

**Comment by t3chnoboy** [comment]
hmm, the ci is failing https://travis-ci.org/t3chnoboy/amazon-product-api/jobs/199458896
`eslint: not found`
I think we need to add `eslint` to `devDependencies`

**Comment by socketwiz** [comment]
Ugh, sorry, let me get that added...

**Comment by socketwiz** [comment]
The older nodes don't seem to like this version of eslint, I'm not sure how you would like to handle that.  I have a fix for the v4 node up, we'll see how that goes.  Its odd how different versions of node find different eslint errors.  I fixed a couple that are not present in my environment but travis somehow uncovered.

**Comment by t3chnoboy** [comment]
Well, I think it's fine to drop the older node versions.
Especially since there's some ongoing work to convert the project to ES6

**Comment by socketwiz** [comment]
OK, all tests passing in travis now...and I learned something about travis, I've never messed with it before.

**Comment by t3chnoboy** [comment]
Awesome!
Do you need this to be published to npm today or can you use git url until we release 1.0?
Will it work for you if I create the `1.0` branch and merge your branch into `1.0`?

There's one more outstanding issue that requires breaking changes and I'd want to include it in 1.0:
https://github.com/t3chnoboy/amazon-product-api/issues/68 
Fixing that issue will close the remaining issues in the milestone
https://github.com/t3chnoboy/amazon-product-api/milestone/1
I'm planning to close out the remaining issues this week. After that we can merge `1.0` into master and publish `1.0.0` to npm.

**Comment by socketwiz** [comment]
> Do you need this to be published to npm today or can you use git url until we release 1.0?

No, the git version is fine for me.

**Comment by socketwiz** [comment]
Thanks for being patient with me, if you guys need anything else done let me know...

**Comment by t3chnoboy** [comment]
@socketwiz thanks for contributing! 😊

**Pull request merged**